Source/Core/Core/CoreTiming.h

```diff
@@ -89,47 +68,113 @@ enum class FromThread
+  // After the first Advance, the slice lengths and the downcount will be reduced whenever an event
+  // is scheduled earlier than the current values (when scheduled from the CPU Thread only).
+  // Scheduling from a callback will not update the downcount until the Advance() completes.
+  void ScheduleEvent(s64 cycles_into_future, EventType* event_type, u64 userdata = 0,
+                     FromThread from = FromThread::CPU);
 
-std::string GetScheduledEventsSummary();
+  // We only permit one event of each type in the queue at a time.
+  void RemoveEvent(EventType* event_type);
+  void RemoveAllEvents(EventType* event_type);
 
-void AdjustEventQueueTimes(u32 new_ppc_clock, u32 old_ppc_clock);
+  // Advance must be called at the beginning of dispatcher loops, not the end. Advance() ends
+  // the previous timing slice and begins the next one, you must Advance from the previous
+  // slice to the current one before executing any cycles. CoreTiming starts in slice -1 so an
+  // Advance() is required to initialize the slice length before the first cycle of emulated
+  // instructions is executed.
+  // NOTE: Advance updates the PowerPC downcount and performs a PPC external exception check.
+  void Advance();
+  void MoveEvents();
 
-u32 GetFakeDecStartValue();
-void SetFakeDecStartValue(u32 val);
-u64 GetFakeDecStartTicks();
-void SetFakeDecStartTicks(u64 val);
-u64 GetFakeTBStartValue();
-void SetFakeTBStartValue(u64 val);
-u64 GetFakeTBStartTicks();
-void SetFakeTBStartTicks(u64 val);
+  // Pretend that the main CPU has executed enough cycles to reach the next event.
+  void Idle();
 
-void ForceExceptionCheck(s64 cycles);
+  // Clear all pending events. This should ONLY be done on exit or state load.
+  void ClearPendingEvents();
+
+  void LogPendingEvents() const;
+
+  std::string GetScheduledEventsSummary() const;
+
+  void AdjustEventQueueTimes(u32 new_ppc_clock, u32 old_ppc_clock);
+
+  u32 GetFakeDecStartValue() const;
+  void SetFakeDecStartValue(u32 val);
+  u64 GetFakeDecStartTicks() const;
+  void SetFakeDecStartTicks(u64 val);
+  u64 GetFakeTBStartValue() const;
+  void SetFakeTBStartValue(u64 val);
+  u64 GetFakeTBStartTicks() const;
+  void SetFakeTBStartTicks(u64 val);
+
+  void ForceExceptionCheck(s64 cycles);
+
+  // Directly accessed by the JIT.
+  Globals& GetGlobals() { return m_globals; }
+
+private:
+  Globals m_globals;
+
+  // unordered_map stores each element separately as a linked list node so pointers to elements
+  // remain stable regardless of rehashes/resizing.
+  std::unordered_map<std::string, EventType> m_event_types;
+
+  // STATE_TO_SAVE
+  // The queue is a min-heap using std::make_heap/push_heap/pop_heap.
+  // We don't use std::priority_queue because we need to be able to serialize, unserialize and
+  // erase arbitrary events (RemoveEvent()) regardless of the queue order. These aren't accomodated
+  // by the standard adaptor class.
+  std::vector<Event> m_event_queue;
+  u64 m_event_fifo_id;
```

Comment:
perhaps primitive type like these should be explicitly given a defaulted initializer, so they're always containing a deterministic value, no matter what stage of initialization core timing is in on construction.

Reply:
Yeah, good idea.